src/requirements/Allowlist/components/AllowlistFormInputs.tsx

```diff
@@ -48,103 +53,194 @@ export default function AllowlistFormInputs({
-    const fileReader = new FileReader()
+  const requirementType = useWatch({ name: `${baseFieldPath}.type` })
 
-    fileReader.onload = () => {
-      const lines = fileReader.result
-        ?.toString()
-        ?.split("\n")
-        ?.filter((line) => !!line)
-        ?.map((line) => line.trim())
+  const uploadFileToGcs = useSubmit(async (file: File) => {
+    const [data, body] = await Promise.all([
+      fetcher(
+        `${process.env.NEXT_PUBLIC_API.replace(
+          "/v1",
+          "/v2"
+        )}/third-party/google/signed-url?requirementType=${
+          requirementType || "ALLOWLIST"
+        }`
+      ),
+      file.arrayBuffer(),
+    ])
 
-      setValue(`${baseFieldPath}.data.addresses`, lines, {
-        shouldValidate: true,
-        shouldDirty: true,
-      })
+    const { signedUrl, fileId } = data || {}
+
+    if (!signedUrl || !fileId) {
+      throw new Error("Failed to upload allowlist")
     }
 
-    fileReader.readAsText(file)
-  }
+    const gcsResponse = await fetch(signedUrl, {
+      method: "PUT",
+      body,
+      headers: {
+        "Content-Type": "text/csv",
+      },
+    })
+
+    if (!gcsResponse.ok) {
+      throw new Error("Failed to upload allowlist")
+    }
+
+    setValue(`${baseFieldPath}.data.addresses`, [], {
+      shouldDirty: true,
+    })
+
+    setValue(`${baseFieldPath}.data.fileId`, fileId, {
+      shouldDirty: true,
+    })
+
+    return fileId
+  })
 
   const { isDragActive, fileRejections, getRootProps, getInputProps } = useDropzone({
     multiple: false,
-    accept: { "text/*": [".csv", ".txt"] },
+    accept: { "text/*": [".csv"] },
     onDrop: (accepted) => {
-      if (accepted.length > 0) parseFile(accepted[0])
+      if (accepted.length > 0) {
+        uploadFileToGcs.onSubmit(accepted[0])
+      }
     },
+    maxSizeMb: 50,
   })
 
   const type = useWatch({ name: `${baseFieldPath}.type` })
 
   return (
     <Stack spacing={4} alignItems="start" {...getRootProps()}>
       <FormControl isInvalid={!!fileRejections?.[0]} textAlign="left">
-        <FormLabel>Upload from file</FormLabel>
+        <FormLabel display={"flex"} gap={2}>
+          Upload from file
+          {addressCount && (
+            <Tooltip
+              label={`Currently an address with ${addressCount} addresses is uploaded. Uploading a new file overwrites these addresses`}
+            >
+              <Tag>{addressCount}</Tag>
+            </Tooltip>
+          )}
+          {uploadFileToGcs.response && (
+            <Button
+              size="xs"
+              variant="ghost"
+              onClick={() => {
+                uploadFileToGcs.reset()
+                setValue(`${baseFieldPath}.data.addresses`, [], {
+                  shouldDirty: true,
+                })
+                setValue(`${baseFieldPath}.data.fileId`, undefined, {
+                  shouldDirty: true,
+                })
+              }}
+            >
+              clear
```

Comment:
Let's use an uppercase `C` here & I think we could also add an icon ;)

Reply:
Good idea! Made some changes in https://github.com/guildxyz/guild.xyz/pull/1244/commits/f69a44165e6801f02af2afd7fc797f7f99002c9a!

Reply:
Nice! Btw I think the `Button` component has an `iconSpacing` prop ;)

Reply:
Nice, thanks, changed to that prop in https://github.com/guildxyz/guild.xyz/pull/1244/commits/bf7dabecf0b8369366828e9a4f136179b8876bdd!